waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim

```diff
@@ -1020,25 +1055,61 @@ proc subscribeToGroupEvents*(ethClientUri: string,
   ## connects to the eth client whose URI is supplied as `ethClientUri`
   ## subscribes to the `MemberRegistered` event emitted from the `MembershipContract` which is available on the supplied `contractAddress`
   ## it collects all the events starting from the given `blockNumber`
-  ## for every received event, it calls the `handler`
+  ## for every received block, it calls the `handler`
   let web3 = await newWeb3(ethClientUri)
-  var latestBlock: Quantity
+  let contract = web3.contractSender(MembershipContract, contractAddress)
+  
+  let blockTableRes = await getHistoricalEvents(ethClientUri, 
+                                                contractAddress, 
+                                                fromBlock=blockNumber)
+  if blockTableRes.isErr():
+    error "failed to retrieve historical events", error=blockTableRes.error()
+    return
+  let blockTable = blockTableRes.get()
+  # Update MT by batch
+  for blockNumber, members in blockTable.pairs():
+    debug "updating the Merkle tree", blockNumber=blockNumber, members=members
+    let res = handler(blockNumber, members)
+    if res.isErr():
+      error "failed to update the Merkle tree", error=res.error()
+
+  # We don't need the block table after this point
+  discard blockTable
+
+  var latestBlock: BlockNumber
+  let handleLog = proc(blockHeader: BlockHeader) {.async, gcsafe.} = 
+    try:
+      let membershipRegistrationLogs = await contract.getJsonLogs(MemberRegistered,
+                                                          blockHash = some(blockheader.hash))
+      if membershipRegistrationLogs.len == 0:
+        return
+      var members: seq[MembershipTuple]
+      for log in membershipRegistrationLogs:
+        let parsedEventRes = parse(MemberRegistered, log)
+        if parsedEventRes.isErr():
+          fatal "failed to parse the MemberRegistered event", error=parsedEventRes.error()
```

Comment:
I believe this error is fatal, since after this point, the waku-rln-relay protocol will stop working correctly.

Reply:
Then we should "panic" out. It should return an error that the receiver will propagate up, and the protocol should interpret and raise as a non-catchable error, a `RlnDefect` (for reference, check the `ResultDefect` in the `nim-stew` repo).

This can be taken into a separate PR; no need to overload this PR with more changes.

Reply:
Will address this in a follow up pr, thanks!